### ISSUE TITLE
Broadcast integral calculation and increase performance

### ISSUE THREAD

**Issue by KirillZubov** [SciML/NeuralPDE.jl#390]
the current calculation of the integral is in for each index in the loop, which is very slow.

https://github.com/SciML/NeuralPDE.jl/blob/2442835ea24fd2e2dbb4663c8d05e69296ac4a50/src/pinns_pde_solve.jl#L870-L873
Need implementation with broadcast for the vector. This will greatly reduce the execution time. 
and will also give the ability to use the GPU for calculation integral.

https://github.com/SciML/NeuralPDE.jl/pull/330#discussion_r686313128

**Comment by killah-t-cell** [comment]
Do you think this is a fast issue? Adding a integral makes the whole thing super slow right now.

how do you think we should implement this?

**Comment by killah-t-cell** [comment]
I played with this a little and actually solved it for 1D. All you have to do is do `return integration_.(cord, lb_, ub_, Ref(flat_θ))` instead of that loop. And then `vcat(...)` a couple of things inside of `integration_`. When we get to multi dimensions though this becomes a bit harder. Currently, a multidimensional integration_ gets 

```julia
cord[:, i] = Float32[0.8, 1.0]
ub__ = [1.0, 1.0]
lb__ = [0.0, 0.0]
```

as arguments. But a broadcast `integration_.(cord, lb_, ub_, Ref(flat_θ))` gives only one dimensional inputs.
```
cord = 0.4f0
lb = 0.0
ub = 1.0
```

I have no idea how to do multidimensional broadcasting. LMK if you have any ideas here @KirillZubov

**Comment by KirillZubov** [comment]
- [ ] minimize the number of unnecessary operations that are called during optimization at each iteration

The next thing that will increase the performance of calculating the integral is to take out all the lines that can be executed once before training. Instead of calling it on each training iteration.

https://github.com/SciML/NeuralPDE.jl/blob/8996411723544bc65e5ee76ff755ee0209b16576/src/pinns_pde_solve.jl#L840-L868

for example:  `typeof(chain)` is static so it can be called one time before
` flat_θ = if (typeof(chain) <: AbstractVector) reduce(vcat,θ) else θ end`

**Comment by killah-t-cell** [comment]
We could just place that line before the map like this, right?

```julia
flat_θ = if (typeof(chain) <: AbstractVector) reduce(vcat,θ) else θ end
return map((cord__,lb__,ub__) -> integration_(cord__,lb__,ub__,flat_θ), eachcol(cord), eachcol(lb_), eachcol(ub_))
```

**Comment by KirillZubov** [comment]
actually, θ should already be flat, so this line is not needed

**Comment by KirillZubov** [comment]
@killah-t-cell  nope, not like this.  Outside initializing the function `integral ` https://github.com/SciML/NeuralPDE.jl/blob/8996411723544bc65e5ee76ff755ee0209b16576/src/pinns_pde_solve.jl#L837

**Comment by killah-t-cell** [comment]
Ah gotcha. But we don't need that line so that is great.

**Comment by killah-t-cell** [comment]
Close issue?

**Comment by killah-t-cell** [comment]
Integration is still incredibly slow for me but I might just need a GPU.

**Comment by ChrisRackauckas** [comment]
Yeah, it will just take more compute. I don't know if there's much of a way around that.

**Comment by killah-t-cell** [comment]
Totally. @ChrisRackauckas would symbolic integration make it faster? Or is that mostly unrelated?

**Comment by ChrisRackauckas** [comment]
Unrelated. It almost certainly won't have a closed form solution and so it will have to be done numerically. This is a quadratic thing, where instead of just evaluating each point in the domain for the PINN loss function, now for each point in the domain you evaluate other points in the domain (for the integral), so you go from O(n) -> O(n^2) in terms of sample points in some sense.

There are fancy things that can be done here though. We could think of doing some kind of scheme for ammortizing the integral cost. I.e., if we integrate from (0,x), then the next point is at (0,y), we really only need to solve the integral on (x,y), but we instead do the whole thing. Also, what about representing the integral itself by a neural network? So if I(...) is approximated by a neural network, it would be cheap to evaluate it at new points in the domain. Thus it would become a two-stage thing in the loss function:

1. Train the integral neural network NN_I to be a good approximation to the integral we want
2. Evaluate the PINN loss

(1) could use a lot of transfer learning: at each step `u(x,y,t)` does not change very much, so `NN_I` would be a decent approximation for the new integral, meaning that we could use the previous weights as the initial condition for the next one, even further decreasing the cost of (1).

Maybe we should capture these into an issue and start executing on some of it.

**Comment by killah-t-cell** [comment]
https://github.com/SciML/NeuralPDE.jl/issues/399 lets move the conversation here then
